Source/Dafny/DafnyAst.cs

```diff
@@ -325,30 +325,67 @@ public class BuiltIns
       return new ArrowType(f.tok, atd, f.Formals.ConvertAll(arg => Resolver.SubstType(arg.Type, typeMap)), Resolver.SubstType(f.ResultType, typeMap));
     }
 
-    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType) {
+    private object MakeTupleKey(List<bool> isGhost, int dims) {
+      if (dims == 0) {
+        return 0;
+      } else {
+        var g = isGhost[dims - 1];
+        return Tuple.Create(g, MakeTupleKey(isGhost, dims - 1));
+      }
+    }
+
+    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType, List<bool> ghostArguments = null) {
       Contract.Requires(tok != null);
       Contract.Requires(0 <= dims);
       Contract.Ensures(Contract.Result<TupleTypeDecl>() != null);
 
       TupleTypeDecl tt;
+      if (ghostArguments == null) {
+        ghostArguments = new bool[dims].ToList().ConvertAll(_ => false);
+      }
+      object key = MakeTupleKey(ghostArguments, dims);
       if (!tupleTypeDecls.TryGetValue(dims, out tt)) {
         Contract.Assume(allowCreationOfNewType);  // the parser should ensure that all needed tuple types exist by the time of resolution
-        if (dims == 2) {
+        if (dims == 2 && !ghostArguments.Contains(true)) {
```

Comment:
I'm not sure, but shouldn't this be `if (dims - ghostArguments.Count(x => x) == 2)` ? I'm guessing the runtime instance of a tuple is the sum of all the non-ghost arguments in the compile-time tuple.

Reply:
Shall we refactor this to
```
var attributes = dims == 2 ? DontCompile() : null;
tt = new TupleTypeDecl(ghostArguments, SystemModule, attributes);
```
?

Reply:
> I'm not sure, but shouldn't this be `if (dims - ghostArguments.Count(x => x) == 2)` ? I'm guessing the runtime size of a tuple is the sum of the non-ghost arguments in the compile-time tuple.

I still have to look at how to compile the ghost components. The branch from which I'm taking most of the work defines different runtime names for tuples that have the same number of concrete components but different number of ghost components, and I'm not sure why.

Reply:
At run time, each n-tuple with m ghost components will use an (n-m)-tuple (except when n-m is 1, in which case I suggest the compiler omit the tuple wrapper altogether). However, I think the enclosing `TupleType` method here needs to generate tuple definitions that _with_ the indicated ghost components.

Reply:
As Runstan wrote, I think we need to declare different tuples for `(Y, Z)` and `(ghost X, Y, Z)`.